### ISSUE TITLE
Fusing a transformer block into one compiled region is ~40% slower on device than splitting attention out

### ISSUE THREAD

**Issue by tdoublep** [torch-spyre/torch-spyre#4746]
## Problem

A BERT-style encoder block compiled as **one** region produces a device kernel ~40% slower than the same block compiled with attention split into its own region: **1689.7 us vs 1181.6 us** for identical math and identical weights. Fusing removes 4 of 5 kernel launches, so it should be faster.

The timing is deterministic (min/max within 0.5% over 8 reps), so this looks like scheduling/codegen rather than a runtime stall. The SDPA primitive survives the fusion — `scaled_dot_product_fused_attention_overrideable` is still in the fused kernel name.

Measured as device kernel time from the AIU profiler. Host wall-clock time dilutes the effect ~3x, because the split variant overlaps much of its device work with host launch latency between kernels.

## Reproducer

Needs `USE_SPYRE_PROFILER=1` (otherwise no `kernel` events).

<details>
<summary>repro_fused_block_device.py</summary>

```python
"""torch-spyre: compiling a transformer block as ONE region is slower on device than
letting attention be its own region.

A BERT-style encoder block (QKV -> SDPA -> out proj -> LayerNorm -> FFN), fp16,
shapes from ibm-granite/granite-embedding-278m-multilingual. Two variants, identical
math and identical weights; only the compiled-region boundary moves:

  split  attention behind an opaque custom op -> several device kernels per block
         (this is what vLLM does today via unified_attention_with_output)
  fused  attention inline                     -> one device kernel per block

Fusing removes 4 of 5 kernel launches, so it should be faster. It is ~20% slower.

Measured as DEVICE KERNEL time from the AIU profiler, not host wall time: host wall
time dilutes the effect ~3x, because the split variant overlaps much of its device
work with host launch latency between kernels. Device kernel time is deterministic
here to <1%, which makes the effect ~100x the noise.

Requires torch-spyre built with USE_SPYRE_PROFILER=1 (otherwise no "kernel" events).
Run:  python repro_fused_block_device.py
"""

import collections
import json
import os
import tempfile

import torch
import torch.nn as nn
import torch.nn.functional as F

B, L, H, D = 4, 512, 12, 64        # batch, seq len, heads, head dim
HID, FF = H * D, 4 * H * D         # 768, 3072
DT = torch.float16
SCALE = D**-0.5
REPS = 8


def attend(qkv, mask):
    q, k, v = qkv.chunk(3, dim=-1)
    q = q.view(B, L, H, D).transpose(1, 2)
    k = k.view(B, L, H, D).transpose(1, 2)
    v = v.view(B, L, H, D).transpose(1, 2)
    a = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, scale=SCALE)
    return a.transpose(1, 2).reshape(B * L, HID)


_attend_compiled = torch.compile(attend, dynamic=False)


@torch.library.custom_op("reprofb::attend", mutates_args={"out"})
def _attend_op(qkv: torch.Tensor, mask: torch.Tensor, out: torch.Tensor) -> None:
    """Opaque to torch.compile, so the block graph must break around it."""
    out.copy_(_attend_compiled(qkv, mask))


@_attend_op.register_fake
def _(qkv, mask, out):
    return None


class Block(nn.Module):
    def __init__(self, fused):
        super().__init__()
        self.fused = fused
        self.qkv = nn.Linear(HID, 3 * HID, dtype=DT)
        self.o = nn.Linear(HID, HID, dtype=DT)
        self.ln1 = nn.LayerNorm(HID, dtype=DT)
        self.fc1 = nn.Linear(HID, FF, dtype=DT)
        self.fc2 = nn.Linear(FF, HID, dtype=DT)
        self.ln2 = nn.LayerNorm(HID, dtype=DT)

    def forward(self, x, mask):
        qkv = self.qkv(x)
        if self.fused:
            a = attend(qkv, mask)
        else:
            a = torch.empty(B * L, HID, dtype=DT, device=x.device)
            torch.ops.reprofb.attend(qkv, mask, a)
        y = self.ln1(self.o(a) + x)
        return self.ln2(self.fc2(F.gelu(self.fc1(y))) + y)


def device_kernels(fused, x, mask):
    """Compile one block, then return {kernel_name: [per-rep device us]} plus output."""
    torch.manual_seed(0)  # identical weights in both variants
    blk = torch.compile(Block(fused).eval().to("spyre"), dynamic=False)

    acts = [torch.profiler.ProfilerActivity.CPU]
    spyre = getattr(torch.profiler.ProfilerActivity, "PrivateUse1", None)
    if spyre is not None:
        acts.append(spyre)

    with torch.inference_mode():
        for _ in range(3):
            out = blk(x, mask)
            out.reshape(-1)[:1].cpu()                 # compile + warm, and sync
        with torch.profiler.profile(activities=acts) as prof:
            for _ in range(REPS):
                out = blk(x, mask)
                out.reshape(-1)[:1].cpu()             # sync, so no async tail leaks
        ref = out.float().cpu()

    path = os.path.join(tempfile.mkdtemp(), "trace.json")
    prof.export_chrome_trace(path)
    kernels = collections.defaultdict(list)
    for e in json.load(open(path))["traceEvents"]:
        if e.get("cat") == "kernel":
            kernels[e["name"]].append(e.get("dur", 0) or 0)
    return kernels, ref


def report(tag, kernels):
    total = sum(sum(v) for v in kernels.values()) / REPS
    count = sum(len(v) for v in kernels.values()) / REPS
    print(f"\n{tag}: {total:8.1f} us/block over {count:.0f} kernel(s)")
    print(f"  {'avg us':>9} {'min':>7} {'max':>7}  kernel")
    for name, ds in sorted(kernels.items(), key=lambda kv: -sum(kv[1])):
        print(f"  {sum(ds) / len(ds):>9.1f} {min(ds):>7.0f} {max(ds):>7.0f}  {name[:74]}")
    return total


def main():
    torch.manual_seed(0)
    x = torch.randn(B * L, HID, dtype=DT).to("spyre")
    kv_lens = torch.tensor([L, 400, 480, 300], dtype=torch.int32).unsqueeze(1)
    pos = torch.arange(L, dtype=torch.int32).unsqueeze(0)
    neg = torch.tensor(torch.finfo(DT).min / 2, dtype=DT)
    mask = torch.where(pos < kv_lens, torch.zeros((), dtype=DT), neg)
    mask = mask.view(B, 1, 1, L).contiguous().to("spyre")

    k_split, out_split = device_kernels(False, x, mask)
    k_fused, out_fused = device_kernels(True, x, mask)

    if not k_split or not k_fused:
        print("NO DEVICE KERNEL EVENTS -- rebuild torch-spyre with USE_SPYRE_PROFILER=1")
        return 1

    print(f"B={B} L={L} heads={H} head_dim={D} hidden={HID} ffn={FF} dtype={DT}")
    print(f"max |split - fused| = {(out_split - out_fused).abs().max().item():.4g}  (same math)")
    t_split = report("split", k_split)
    t_fused = report("fused", k_fused)

    copy = sum(sum(v) for k, v in k_split.items() if "copy_from_d2d" in k) / REPS
    print(f"\nfused / split = {t_fused / t_split:.3f}x  ({t_fused - t_split:+.1f} us/block)")
    if copy:
        print(f"  split includes {copy:.1f} us of out.copy_ that the vLLM path does not have;")
        print(f"  excluding it, fused / split = {t_fused / (t_split - copy):.3f}x")
    print("\nfused has 1/5 the kernel launches, so it should be faster, not slower.")
    return 0


raise SystemExit(main())
```

</details>

## Output

```
B=4 L=512 heads=12 head_dim=64 hidden=768 ffn=3072 dtype=torch.float16
max |split - fused| = 0.01562  (same math)

split:   1181.6 us/block over 5 kernel(s)
     avg us     min     max  kernel
      455.0     451     464  spyre_kernel_v1_fused_add_gelu_layer_norm_linear_44fr32g372xmmbzh#2
      437.2     430     442  spyre_kernel_v1_fused_scaled_dot_product_fused_attention_overrideable_spli
      207.4     204     213  spyre_kernel_v1_fused_linear_nwtdha3ipkmsmewd#2
       59.8      59      61  spyre_kernel_v1_fused_copy_from_d2d_pset5xkbe4zhd2uc#2
       22.3      22      23  spyre_kernel_v1_fused_unknown_onmt24hvwl4f5xys#2

fused:   1689.7 us/block over 1 kernel(s)
     avg us     min     max  kernel
     1689.7    1686    1694  spyre_kernel_v1_fused_add_gelu_layer_norm_linear_scaled_dot_product_fused_

fused / split = 1.430x  (+508.0 us/block)
  split includes 59.8 us of out.copy_ that the vLLM path does not have;
  excluding it, fused / split = 1.506x
```

The same numbers appear end-to-end in vLLM on `ibm-granite/granite-embedding-278m-multilingual`: 1680.3 us/layer fused vs 1200.0 us/layer split, which is ~7.6% throughput loss over 12 layers.

## Environment

- torch-spyre `ef2dd27f781e7e3711e796db079819f4499377d5`, built with `USE_SPYRE_PROFILER=1`
- torch `2.13.0+cpu`
- RHEL 10.2, x86_64
- `ibm-deeptools` 2.0.0-0.main.1+2457.d868b05_349
- `ibm-flex` 2.0.0-0.main.1+570.c11584a_408
- `ibm-senlib-core` 2.0.0-0.main.1+275.7b07aaf_268
- `ibm-spyre-comms` 1.0.0-0.main.1+154.ce55e90_197
- `ibm-libaiupti` 2.0.0-0.main.1+30.ef4e622_23
